lib/spack/spack/spec.py

```diff
@@ -1732,18 +1781,22 @@ def read_yaml_dep_specs(dependency_dict):
         for dep_name, elt in dependency_dict.items():
             if isinstance(elt, six.string_types):
                 # original format, elt is just the dependency hash.
-                dag_hash, deptypes = elt, ['build', 'link']
+                dag_hash, deptypes, virtuals = elt, ['build', 'link'], []
             elif isinstance(elt, tuple):
                 # original deptypes format: (used tuples, not future-proof)
                 dag_hash, deptypes = elt
+                virtuals = []
             elif isinstance(elt, dict):
                 # new format: elements of dependency spec are keyed.
                 dag_hash, deptypes = elt['hash'], elt['type']
+                # {}.get instead of subscripting is needed for backward
+                # compatibility. Without it reindex would fail
+                virtuals = elt.get('provides', [])
```

Comment:
> I think this means you can iterate over the dependency conditions in self.package.dependencies and test them against self, and see which virtuals self actually depended on. The ones that overlap with what the dependency provides are what should go here

Not clear to me how we can do that here without a reworking of the method. The code is in a `staticmethod` and we have no `self` in this context. I can add a check at the end of `Spec.from_dict` and reconstruct `provides` on edges there, but that check will be run for every spec read from a YAML file regardless if it comes from an old DB or not.

Reply:
I'll try using `None` as a signal to `Spec.from_dict` to say that the old format was encountered.

Reply:
@tgamblin Done in 22c11c7. Don't know if it's worth keeping the commit, since it spreads logic to treat the old DB in multiple places. Let me know if you want some change on that or discard it completely.